stan/math/rev/mat/functor/algebra_solver.hpp

```diff
@@ -0,0 +1,246 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_ALGEBRA_SOLVER_HPP
+
+#include <stan/math/rev/mat/functor/algebra_system.hpp>
+#include <stan/math/prim/mat/fun/mdivide_left.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/scal/meta/is_var.hpp>
+#include <stan/math/prim/scal/err/check_finite.hpp>
+#include <stan/math/prim/scal/err/check_consistent_size.hpp>
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <unsupported/Eigen/NonLinearOptimization>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * The vari class for the algebraic solver. We compute the  Jacobian of
+     * the solutions with respect to the parameters using the implicit
+     * function theorem. The call to Jacobian() occurs outside the call to
+     * chain() -- this prevents malloc issues.
+     *
+     * Members:
+     * y_ the vector of parameters
+     * y_size_ the number of parameters
+     * x_size_ the number of unknowns
+     * theta_ the vector of solution
+     * Jx_y_ the Jacobian of the solution with respect to the parameters.
+     */
+    template <typename FS, typename F, typename T, typename FX>
+    struct algebra_solver_vari : public vari {
+      vari** y_;
+      int y_size_;
+      int x_size_;
+      vari** theta_;
+      Eigen::MatrixXd Jx_y_;
+
+      algebra_solver_vari(const FS& fs,
+                          const F& f,
+                          const Eigen::VectorXd x,
+                          const Eigen::Matrix<T, Eigen::Dynamic, 1> y,
+                          const std::vector<double> dat,
+                          const std::vector<int> dat_int,
+                          const Eigen::VectorXd theta_dbl,
+                          FX& fx,
```

Comment:
The function `get_jacobian()` gets called on `FX`. Now, in `algebra_system.hpp`:
```
 Eigen::MatrixXd get_jacobian(const Eigen::VectorXd &dv) {
        Eigen::VectorXd fvec;
        jacobian(fs_, dv, fvec, J_);
        return J_;
      }
```
and looking at the signature of jacobian: `jacobian(F&f, Matrix& x, Matrix& fx, Matrix& J)`. So, the third argument is passed as a non-constant reference, which is why `FX` cannot be a `const` in the first place. If we wanted `FX` to be constant (for safety measures), we would need to create in `get_jacobian` a copy of `FX` which gets passed to `jacobian`. So there is a trade-off.

I'm ok with how it currently is, but you may think we should pick safety over a (minor?) efficiency gain.